### PR TITLE
tests: kernel: userspace: Add some more tests to userspace protection tests

### PR DESCRIPTION
**Pull request by spoorthik**
Adding few more test scenarios to validate user space protection. 

start_essential_thread: User thread cannot create essential thread.
high_prio_uthread: User thread can create a thread of equal or lower priority.
thread_in_multiple_domains: A thread can belong to single memory domain at a time.
add_thread: Cannot add a thread to a destroyed memory domain.
check_access_perms: Threads belong to same memory domain will have same access permissions to its partitions.

Signed-off-by: Spoorthi K <spoorthi.k@intel.com>

### CODE REVIEW

File: tests/kernel/mem_protect/userspace/src/main.c
Comment (andrewboie) [inline tests/kernel/mem_protect/userspace/src/main.c]: I think it's worth noting that entering user mode, when you are already in user mode, should not have specific runtime checks for it. Some platforms may crash, it's really undefined behavior. I'd like to see this removed and at most, checked in the kernel with an _ASSERT

File: tests/kernel/mem_protect/userspace/src/main.c
Comment (andrewboie) [inline tests/kernel/mem_protect/userspace/src/main.c]: why is this here? remove

File: tests/kernel/mem_protect/userspace/src/main.c
Comment (andrewboie) [inline tests/kernel/mem_protect/userspace/src/main.c]: why is this here?

File: tests/kernel/mem_protect/userspace/src/main.c
Comment (andrewboie) [inline tests/kernel/mem_protect/userspace/src/main.c]: why is this here?

**Review by tyrantlizard77 — DISMISSED** [review]
CI is really freaking out, not sure if related to this patch, but need a passing CI result for review on this to continue.

File: tests/kernel/mem_protect/userspace/src/main.c
Comment (stephensmalley) [inline tests/kernel/mem_protect/userspace/src/main.c]: Same problem as above - you can't use k_mem_domain_add_thread() from usermode currently AFAICT.  We are faulting on the irq_lock() but not due to use of a destroyed memory domain.  And we'd only hit the assertion if assertions were enabled in the config.

**Review by stephensmalley — CHANGES_REQUESTED** [review]

File: tests/kernel/mem_protect/userspace/src/main.c
Comment (stephensmalley) [reply tests/kernel/mem_protect/userspace/src/main.c]: Actually, you need to k_thread_abort(k_current_get()).  Otherwise this function will return back through its call chain to test_cb(), set test_result to 0, and call k_sem_give(&test_end_signal), waking run_test() and indicating test success irrespective of the actual test result.  This is a problem whenever creating a separate thread from one of these test functions and performing the ztest_test_fail/pass from the other thread; you need to abort the current one to prevent incorrect reporting.

File: tests/kernel/mem_protect/userspace/prj.conf
Comment (stephensmalley) [inline tests/kernel/mem_protect/userspace/prj.conf]: Do we want this enabled for these tests?

File: tests/kernel/mem_protect/userspace/src/main.c
Comment (stephensmalley) [inline tests/kernel/mem_protect/userspace/src/main.c]: Why K_PRIO_PREEMPT()?

File: tests/kernel/mem_protect/userspace/src/main.c
Comment (stephensmalley) [inline tests/kernel/mem_protect/userspace/src/main.c]: This is not a test of userspace AFAICT; it is testing kernel assertion checking on kernel threads performing memory domain setup.  Probably belongs in a separate test program? Also, I am not a fan of this k_sem_take(&sync_sem, SEM_TIMEOUT) idiom for tests; it seems unreliable (could timeout and yield a false pass result).  We are using the approach k_thread_abort(k_current_get()) in the ztest thread to stop the ztest thread from generating a pass or fail, and instead handling both in the newly created thread.

Comment (stephensmalley) [reply tests/kernel/mem_protect/userspace/src/main.c]: See revoke_other_thread() for an example of this approach.

File: tests/kernel/mem_protect/userspace/src/main.c
Comment (stephensmalley) [inline tests/kernel/mem_protect/userspace/src/main.c]: Same problem as in https://github.com/zephyrproject-rtos/zephyr/pull/5197#discussion_r166453992

**Review by stephensmalley — DISMISSED** [review]
See prior comments, plus something went wrong in your merge/rebase - it doesn't compile.
/home/sds/zephyr-project/tests/kernel/mem_protect/userspace/src/main.c: In function ‘destroy_mem_domain’:
/home/sds/zephyr-project/tests/kernel/mem_protect/userspace/src/main.c:669:19: warning: missing terminating " character
     &app0_domain, "The thread has reference to
                   ^
/home/sds/zephyr-project/tests/kernel/mem_protect/userspace/src/main.c:670:45: warning: missing terminating " character
     memory domain which is already destroyed");
                                             ^
...

File: tests/kernel/mem_protect/userspace/prj.conf
Comment (spoorthik) [reply tests/kernel/mem_protect/userspace/prj.conf]: CONFIG_ASSERT is enabled for the test thread_in_multiple_domains(), which  checks for assert condition in k_mem_domain_add() API. I agree that the test should belong to set of tests related to memory domains in #5197. I will move this test once #5197 gets merged.

Comment (nashif) [reply tests/kernel/mem_protect/userspace/prj.conf]: CONFIG_ASSERT is enabled by default with sanitycheck

Comment (spoorthik) [reply tests/kernel/mem_protect/userspace/prj.conf]: @nashif But to execute the thread_in_multiple_domains() locally, we may need this. The test is meant to hit assert failure condition. If we do not enable it, the test case will fail without checking for assertion when run locally.

File: tests/kernel/mem_protect/userspace/src/main.c
Comment (spoorthik) [reply tests/kernel/mem_protect/userspace/src/main.c]: Yes, I agree that this belong to separate set of tests. This test case will be merged with #5197. In this test, ASSERT failure is hit and the thread goes to infinite loop. Hence calling k_sem_take(&sync_sem, SEM_TIMEOUT) in the main thread. I will add a condition to check from ztest to explicitly know which part of the ASSERT is failed.

**Comment by spoorthik** [comment]
@andrewboie @stephensmalley  @lpereira : Thanks for the review.  I have addressed the review comments.

File: tests/kernel/mem_protect/userspace/prj.conf
Comment (andrewboie) [reply tests/kernel/mem_protect/userspace/prj.conf]: The problem with CONFIG_ASSERT is that when an assertion hits, it just hangs the system instead of having ztest report success or failure.
Is there any other way we can detect this situation in the test itself and fail it accordingly?

File: tests/kernel/mem_protect/userspace/src/main.c
Comment (stephensmalley) [inline tests/kernel/mem_protect/userspace/src/main.c]: The memory domain tests all appear to fail on ARM w/ current master (which has pr-4974 merged), with:
ASSERTION FAIL [writable != executable] @ /home/sds/zephyr-project/kernel/mem_domain.c:21:
        writable page not executable

Also I still have doubts about whether these belong here, particularly the ones that don't create any userspace thread.

Comment (spoorthik) [reply tests/kernel/mem_protect/userspace/src/main.c]: These tests are generic, I don't think the failure is because of the test case.  I will merge this test case with #5197 once it is merged. It is more related to those set of test cases.

File: tests/kernel/mem_protect/userspace/prj.conf
Comment (spoorthik) [reply tests/kernel/mem_protect/userspace/prj.conf]: I agree. But here in this test, ztest thread is creating another thread and this thread hits assertion. The ztest thread then checks if the assert condition is hit and reports the test result.

**Comment by spoorthik** [comment]
#5197 is merged, will revisit the merged test cases and create pull requests to add new set of test cases.

**Pull request closed without merging**
